### DISCUSSION TITLE
access the field, value and operator of request.criteria inside the endpoint

### DISCUSSION THREAD

**Discussion by teafmartins** [serenity-is/Serenity#6971] [Help]
Hi,
I have a modified endpoint. Within the endpoint, I need access to all the fields used in the filter criteria, value and operator (request.Criteria). 
How can I get the field, value and operator used in the criteria?

I was trying something like this, but I get this error.
![image](https://github.com/serenity-is/Serenity/assets/17210232/f1a689de-6277-4459-bfee-966ebec50e29)

Thx

**Comment by edwardch** [comment]
I haven't had to do this myself, but checking in the comments of Serenity.Services.ListRequest, the Criteria is `The where criteria for the query. This is passed as an array of arrays in the JSON.`

You may want to try put a breakpoint in your endpoint and add a watch to your request.Criteria, then work from there. I'm curious to know what the solution may look like.

**Comment by VictorTomaili** [comment]
Check the [JsonCriteriaConverter](https://github.com/serenity-is/Serenity/blob/master/src/Serenity.Net.Data/Criteria/JsonCriteriaConverter.cs)

**Comment by teafmartins** [comment]
@VictorTomaili , Right, I think that's the way to go. But I'm having trouble coming up with an example... is it too much to ask for a small one on how to use it?

Thank you

**Reply by VictorTomaili** [reply]
There is no example except the JsonCriteriaConverter which I know.

**Comment by volkanceylan** [comment]
See https://github.com/serenity-is/Serenity/blob/master/src/Serenity.Net.Data/Criteria/SafeCriteriaValidator.cs and https://github.com/serenity-is/Serenity/blob/master/src/Serenity.Net.Data/Criteria/BaseCriteriaVisitor.cs

**Comment by volkanceylan** [comment]
Also https://github.com/serenity-is/Serenity/blob/master/src/Serenity.Net.Services/RequestHandlers/Helpers/CriteriaFieldExpressionReplacer.cs

You need to know how to work on a tree. Criteria is a binary tree with AND/OR operators etc., so it is not as simple as enumerating some list.
